### PR TITLE
Fix empty host list for instance stoppable response

### LINKED ISSUE #1232: Stoppable rest response is empty because of instance check exception
**Describe the bug**
Health check stoppable responds with empty hostname

**To Reproduce**
Turn off the cluster config: "PERSIST_INTERMEDIATE_ASSIGNMENT": "false"
Issue REST request: `/namespaces/<namespace>/clusters/<cluster>/instances?command=stoppable -d ' {"instances": ["instance"], "selection_base": "zone_based", "max_instance": "2", "customized_values": "{}"}' -H "Content-Type: application/json"`

Response has empty contents:
```
{
"instance_stoppable_parallel" : [ ],
"instance_not_stoppable_with_reasons" : { }
}
```

**Expected behavior**
```
{
"instance_stoppable_parallel" : [ "instance" ],
"instance_not_stoppable_with_reasons" : { }
}
```
**Additional context**
"PERSIST_INTERMEDIATE_ASSIGNMENT" needs to be turned on so REST API could check the instance is already in stable state: IS matches EV, as a condition of an instance stoppable.

### PR DESCRIPTION
**Pull request by pkuwm**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixed #1232 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Health check stoppable responds with empty hostname.

This PR adds error message of instance check to API response so users have a clear idea of the reason if PERSIST_INTERMEDIATE_ASSIGNMENT is not turned on.
```
curl -XPOST http://localhost:8100/admin/v2/namespaces/namespacce/clusters/cluster/instances\?command\=stoppable -d ' {"instances": ["instance"], "selection_base": "zone_based", "max_instance": "2", "customized_values": "{}"}' -H "Content-Type: application/json"
{
  "instance_stoppable_parallel" : [ ],
  "instance_not_stoppable_with_reasons" : {
    "instance" : [ "HELIX:INVALID_CONFIG" ]
  }
}
```

Server log will have the reason logged: `Cluster config PERSIST_INTERMEDIATE_ASSIGNMENT is not turned on, which is required for instance stability check.`

### Tests

- [x] The following tests are written for this issue:

 - testBatchGetInstancesStoppableChecksWhenException()

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Tests run: 166, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 163.995 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 166, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:49 min
[INFO] Finished at: 2020-08-07T21:27:36-07:00
[INFO] ------------------------------------------------------------------------
```
### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Comment by pkuwm** [comment]
> Error message for server side is clear to find out the fail reason. The more important is let user to understand what is the real fail reason. Do we have that if persist assignment turned off, at least, we gave user info EV and IS does not match.

@dasahcc This PR is to provide the info to let users know the reason. The root cause is exception/error is not returned as response. This PR adds the exception/error message to not stoppable reason in the return response.
https://github.com/apache/helix/pull/1237/files#diff-c98c0908866d29b64f625e799c198d15R224-R227

The result will be what is the description
```
curl -XPOST http://localhost:8100/admin/v2/namespaces/namespacce/clusters/cluster/instances\?command\=stoppable -d ' {"instances": ["instance"], "selection_base": "zone_based", "max_instance": "2", "customized_values": "{}"}' -H "Content-Type: application/json"
{
  "instance_stoppable_parallel" : [ ],
  "instance_not_stoppable_with_reasons" : {
    "instance" : [ "HELIX:Instance stability check needs persist assignment cluster config turned on: PERSIST_INTERMEDIATE_ASSIGNMENT = true" ]
  }
}
```

**Comment by dasahcc** [comment]
> > Error message for server side is clear to find out the fail reason. The more important is let user to understand what is the real fail reason. Do we have that if persist assignment turned off, at least, we gave user info EV and IS does not match.
> 
> @dasahcc This PR is to provide the info to let users know the reason. The root cause is exception/error is not returned as response. This PR adds the exception/error message to not stoppable reason in the return response.
> https://github.com/apache/helix/pull/1237/files#diff-c98c0908866d29b64f625e799c198d15R224-R227
> 
> The result will be what is the description
> 
> ```
> curl -XPOST http://localhost:8100/admin/v2/namespaces/namespacce/clusters/cluster/instances\?command\=stoppable -d ' {"instances": ["instance"], "selection_base": "zone_based", "max_instance": "2", "customized_values": "{}"}' -H "Content-Type: application/json"
> {
>   "instance_stoppable_parallel" : [ ],
>   "instance_not_stoppable_with_reasons" : {
>     "instance" : [ "HELIX:Instance stability check needs persist assignment cluster config turned on: PERSIST_INTERMEDIATE_ASSIGNMENT = true" ]
>   }
> }
> ```

File: helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java [hunk 1]
```diff
@@ -213,7 +219,12 @@ public StoppableCheck getInstanceStoppableCheck(String clusterId, String instanc
           instancesForNextCheck.add(instance);
         }
       } catch (InterruptedException | ExecutionException e) {
-        LOG.error("Failed to get StoppableChecks in parallel. Instance: {}", instance, e);
+        LOG.warn("Failed to get StoppableChecks in parallel. Instance: {}. Caused by {}", instance,
```
Comment (jiajunwang) [inline helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java:222]: Why it is a warning now?

Comment (jiajunwang) [reply helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java]: I think the failure reason does not really matter here (although an Exception looks to be a problem to me more than a warning).
The real matter thing is that whether the result is complete and fully computed. If not certain, then this is an error.

Comment (pkuwm) [reply helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java]: The code change is not necessary and kept as it is.

Regarding warn or error, I look at it from server perspective: if the exception is not a server exception that affects running of the REST server, I would say it is an error.
But obviously this is not an exception that stops REST server running. It is just a normal running of the API check. It should not be logged as an error. A warning should already be enough to help debug.

**Comment by pkuwm** [comment]
Thanks, @dasahcc @jiajunwang, for the review.

This PR is ready to be merged, approved by @dasahcc

**Pull request merged**